presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionarySelectiveStreamReader.java

```diff
@@ -275,15 +275,9 @@ private void openRowGroup()
     {
         // read the dictionary
         if (!dictionaryOpen && dictionarySize > 0) {
-            if (dictionary.length < dictionarySize) {
-                dictionary = new long[dictionarySize];
-            }
-
-            LongInputStream dictionaryStream = dictionaryDataStreamSource.openStream();
-            if (dictionaryStream == null) {
-                throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Dictionary is not empty but data stream is not present");
-            }
-            dictionaryStream.nextLongVector(dictionarySize, dictionary);
+            GetDictionaryResult getDictionaryResult = dictionaryProvider.getDictionary(streamDescriptor, dictionary, dictionarySize);
+            dictionary = getDictionaryResult.getDictionaryBuffer();
+            isDictionaryOwner = getDictionaryResult.getIsBufferOwner();
```

Comment:
After reading this logic, do we see a flaw here? If the current stream reader is the dictionary reader for the previous stripe. Then, this logic will drop the reference making the previous dictionary's memory is not tracked by any stream reader. If all stream readers drop the reference on that, then yes, we will GC the dictionary. But is that going to be the case. Seems like a reference counting method is safer on the memory accounting?

cc: @arunthirupathi

Reply:
There are two cases here.

1. Not flat map dictionary ( all existing use case). Here there is only one owner, so the code path will always be counted.
2. Shared flat map dictionary ( new use case). In this case, in the next stripe all readers are allocated new, and only one of them will be the owner. 
So there should not be any accounting bug.

Also in my opinion the lifetime of the dictionary and the memory accounting are independent. JVM ensures the lifetime. This change could introduce bugs in the memory accounting (I don't think there is one here).

Reply:
Yes, case 2 is the one I'm referring to. Suppose we have two stream readers R1 and R2. R1 is currently the dictionary owner of stripe S1. Then the memory assumption is R1's retained bytes + R2's retained bytes + S1's dictionary bytes. Now suppose R1 is going to open the next row group. R1 will drop the ownership of S1 and start to count S2's dictionary memory. Now, the memory accounting will be R1's retained bytes + R2's retained bytes + S2's dictionary bytes. However, R1's dictionary is actually still held by R2 as R2 has not entered the new stripe yet. So the correct accounting would be R1's retained bytes + R2's retained bytes + S1's dictionary bytes + S2's dictionary bytes. Once R2 drops the references of S1's dictionary, we can safely remove S1's accounting. That is why I mentioned the reference count approach, which is closer to JVM GC.

Regarding "lifetime of the dictionary and the memory accounting are independent", the purpose of memory accounting is to estimate as perfect as possible the actual JVM memory usage for the current thread/task/query. Though it won't be exactly accurate. We usually count an object's memory usage once it's created and subtract the memory usage once we know it will be fully dereferenced. For example an object is set to null or go out of a scope. In such case, the object will not be reachable in JVM reference graph.

Coming back to this case, if we do believe that the interval of switching from S1 to S2 is short, then likely the current approach is ok. But a comment would be great. Something like "we drop the ownership of a dictionary and creating a new one with the assumption that all other stream readers will drop the reference on it in a fairly short amount of time."

Reply:
In case 2, R1 and R2 will be freed at the end of the Stripe S2. For Stripe S2 we create two more readers R3 and R4. 

https://github.com/prestodb/presto/blob/master/presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatSelectiveStreamReader.java#L638

So the memory accounting should be right. By the way, love your detailed comment with the examples.

Reply:
I believe this case is going to be ok because by the time R2 opens dictionary in S2, previous R1 is already "destroyed" and waiting to be collected. We shouldn't need to worry about the same R1 dropping ownership of dictionary. There should be no reference to any dictionaries in S1 by the time R2 opens dictionary in S2. So we are really looking at the delay of GC here.

Reply:
@HuamengJiang  It warrants a comment of the discussion above.